src/feature/hs/hs_client.c

```diff
@@ -956,6 +956,27 @@ client_get_random_intro(const ed25519_public_key_t *service_pk)
   return ei;
 }
 
+/** Called when a rendezvous circuit has timed out. Every streams attached to
+ * the circuit will get set with the SOCKS5_HS_REND_FAILED (0xF3) extended
+ * error code so if the connection to the rendezvous point ends up not
+ * working, this code could be sent back as a reason. */
+static void
+socks_report_rend_circuit_timed_out(const origin_circuit_t *rend_circ)
+{
+  tor_assert(rend_circ);
+
+  /* For each entry connections attached to this rendezvous circuit, report
+   * the error. */
+  for (edge_connection_t *edge = rend_circ->p_streams; edge;
+       edge = edge->next_stream) {
```

Comment:
Could this circuit have no streams? Do we care? Will this crash?

Reply:
I believe it is not possible. If the SOCKS connection is closed, before we remove the stream from the circuit, we would close the circuit and thus updating the error code and then closing the SOCKs connection.

If it ever happens, won't matter, the loop will just be NOP.